src/grammar/expressions.rs

```diff
@@ -90,61 +113,134 @@ fn expr_bp(p: &mut Parser, min_bp: u8) {
             continue;
         }
 
-        if let Some(((l_bp, r_bp), syntax_kind)) = infix_bp(op) {
-            if l_bp < min_bp {
+        if let Some(operator) = infix_bp(op) {
+            if operator.bp.l_bp() < min_bp {
                 break;
             }
 
-            match syntax_kind {
+            match operator.mapping {
                 Some(syntax_kind) => p.bump_any_map(syntax_kind),
                 None => p.bump_any(),
             }
 
-            add_expr_node(p, checkpoint, Some(r_bp));
+            if let Some(cb) = operator.callback {
+                cb(p, operator.bp.r_bp());
+            }
+
+            add_expr_node(p, checkpoint, Some(operator.bp.r_bp()));
             continue;
         }
 
         break;
     }
+    Ok(())
 }
 
 fn add_expr_node(p: &mut Parser, checkpoint: Checkpoint, sub_expr: Option<u8>) {
     p.start_node_at(checkpoint, SyntaxKind::Expression);
 
     if let Some(min_bp) = sub_expr {
-        expr_bp(p, min_bp)
+        let _ = expr_bp(p, min_bp);
     };
 
     p.finish();
 }
 
-fn prefix_bp(op: TokenKind) -> Option<(u8, Option<SyntaxKind>)> {
-    match op {
-        TokenKind::NotKw => Some((5, Some(SyntaxKind::LogicOp))),
-        TokenKind::Plus | TokenKind::Minus => Some((15, None)),
-        _ => None,
+struct BindingPower(u8, u8);
```

Comment:
At first I did not understand the intention of this, would it be more appropriate to consider this a precedence level rather than a binding power?

When I check the Rust Analyzer code for reference there are only a few initial values for a (single) binding power value, for example in the [expr](https://github.com/rust-lang/rust-analyzer/blob/master/crates/parser/src/grammar/expressions.rs#L21) parse function. It's not really meant for operator precedence, at least that's not what I found but I might have missed something. The binding power is also increased for inner blocks (see [here](https://github.com/rust-lang/rust-analyzer/blob/master/crates/parser/src/grammar/expressions.rs#L309)).

Is it necessary to have two values, left and right binding power or does that basically mimic the `+1` behaviour? Otherwise all given pairs are consecutive numbers of pairs, e.g. `1, 2`, `3, 4`.

Reply:
I initially followed these two Pratt parser articles:
* https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html
* https://arzg.github.io/lang/10/

None of them state the actual reason to provide the right binding power, but my guess is for operators such as exponentials: `2^3^4` should equal `(2^(3^4))` and not `((2^3)^4)`. For this to actually happen in our parser, the `r_bp` of the exponential operator would have to lower than the `l_bp`.

That being said, all common RDBMSs know to me provide some variation of a `POWER()` function.

What do you think, should we remove them to simplify the code, or keep them in case we need to cover an operator such as `^` in the future?

Reply:
Ok, to amend, Rust Analyzer uses a separate function [`current_op`](https://github.com/rust-lang/rust-analyzer/blob/master/crates/parser/src/grammar/expressions.rs#L195) to determine precedence of the current operator. The code above may not necessarily follow the exact same logic, but is applying precedence with an infix operator.

Reply:
I would suggest to simplify the code for now. If that specific use case becomes relevant it's probably a good idea to handle that explicitely.